src/connection/conn_string.rs

```diff
@@ -0,0 +1,432 @@
+    pub charset: Option<Charset>,
+    pub dialect: Option<Dialect>,
+    pub lib_path: Option<String>,
+    pub stmt_cache_size: Option<usize>,
+}
+
+/// Parse the connection string.
+///
+/// Basic string sintax: `firebird://{user}:{pass}@{host}:{port}/{db_name}?{options}`
+pub fn parse<S: Into<String>>(conn_s: S) -> Result<ConnStringSettings, FbError> {
+    let sconn = conn_s.into();
+
+    if !sconn.starts_with("firebird://") {
+        return Err(FbError::from(
+            "The string must start with the prefix 'firebird://'",
+        ));
+    }
+
+    let user = regex_find(r#"(?:(/))([[:alnum:]]+)(?:.*)(?:@)"#, &sconn, 2, false)?;
+
+    let pass = regex_find(r#"(?:(:))([[:alnum:]]+)(?:(@))"#, &sconn, 2, false)?;
+
+    let host = regex_find(
+        r#"((?:://)|(?:@))([^@/:]+)((?:\w:/)|(?::[[:digit:]])|(?:/))"#,
+        &sconn,
+        2,
+        true,
+    )?;
+
+    let port = {
+        let fport_op = regex_find(r#"(?:(:))([[:digit:]]+)(?:(/))"#, &sconn, 2, true)?;
+        if let Some(fport) = fport_op {
+            match fport.parse::<u16>() {
+                Ok(v) => Some(v),
+                _ => None,
+            }
+        } else {
+            None
+        }
+    };
+
+    let db_name = {
+        // remote host
+        let mut db_name = regex_find(r#"((?:@\w+/)|(?:[0-9]/))([^\?]+)"#, &sconn, 2, true)?;
+        if db_name.is_none() {
+            // embedded
+            db_name = regex_find(r#"(?://)([^\?]+)"#, &sconn, 1, true)?;
+        }
+
+        db_name.ok_or_else(|| FbError::from("The database name/path is required"))?
+    };
+
+    let lib_path = regex_find(r#"(?:\?)(?:.*)(lib=)([^&]+)"#, &sconn, 2, false)?;
+
+    let dialect = {
+        let fdialect_op = regex_find(r#"(?:\?)(?:.*)(dialect=)([[:digit:]])"#, &sconn, 2, false)?;
+        if let Some(fdialect) = fdialect_op {
+            match Dialect::from_str(&fdialect) {
+                Ok(d) => Some(d),
+                _ => None,
+            }
+        } else {
+            None
+        }
+    };
+
+    let charset = {
+        let fcharset_op = regex_find(r#"(?:\?)(?:.*)(charset=)([^&]+)"#, &sconn, 2, false)?;
+        if let Some(fcharset) = fcharset_op {
+            match Charset::from_str(&fcharset) {
+                Ok(d) => Some(d),
+                _ => None,
+            }
+        } else {
+            None
+        }
+    };
+
+    let stmt_cache_size = {
+        let fstmt_op = regex_find(
+            r#"(?:\?)(?:.*)(stmt_cache_size=)([[:digit:]]+)"#,
+            &sconn,
+            2,
+            true,
+        )?;
+        if let Some(fstmt) = fstmt_op {
+            match fstmt.parse::<usize>() {
+                Ok(v) => Some(v),
+                _ => None,
+            }
+        } else {
+            None
+        }
+    };
+
+    Ok(ConnStringSettings {
+        user,
+        pass,
+        host,
+        port,
+        db_name,
+        charset,
+        dialect,
+        lib_path,
+        stmt_cache_size,
+    })
+}
```

Comment:
@fernandobatels

I have extracted each regex into a `const &str` here: [link](https://github.com/afsmi/rsfbclient/blob/connstr_regex_tweaks/src/connection/conn_string.rs#L31-L40)

to save you a click I will copy it here:
```
    const RGX_USER: &str = r#"(?:(/))([[:alnum:]]+)(?:.*)(?:@)"#;
    const RGX_PASS: &str = r#"(?:(:))([[:alnum:]]+)(?:(@))"#;
    const RGX_HOST: &str = r#"((?:://)|(?:@))([^@/:]+)((?:\w:/)|(?::[[:digit:]])|(?:/))"#;
    const RGX_PORT: &str = r#"(?:(:))([[:digit:]]+)(?:(/))"#;
    const RGX_DB_NAME: &str = r#"((?:@\w+/)|(?:[0-9]/))([^\?]+)"#;
    const RGX_DB_NAME_EMBEDDED: &str = r#"(?://)([^\?]+)"#;
    const RGX_LIBPATH: &str = r#"(?:\?)(?:.*)(lib=)([^&]+)"#;
    const RGX_DIALECT: &str = r#"(?:\?)(?:.*)(dialect=)([[:digit:]])"#;
    const RGX_CHARSET: &str = r#"(?:\?)(?:.*)(charset=)([^&]+)"#;
    const RGX_STMT_CACHE_SIZE: &str = r#"(?:\?)(?:.*)(stmt_cache_size=)([[:digit:]]+)"#;
```



Could you provide some explanations (either via adding inline code comments, or here in the PR ):
* intended behavior of each regex, including:
  * the purpose of non-obvious capture groups (use your judgment), for example:
    * in `RGX_HOST`: why is there a `://` group at the beginning? is it to handle hosts of the form `uri_scheme://...` ?
    * in `RGX_DB_NAME`: what is the intention of `(?:@\w+/)|(?:[0-9]/))` at the beginning? I think it is to start matching only after the host, but the intention isn't really clear
  * a list of format variations that are intended to be handled
* which combinations of sometimes-optional parameters should be allowed and disallowed at the parser level (ignoring any validation that may be done later).

also: i believe usernames and passwords are allowed to contain special characters including `@` and `:`, though in the case of using `isql` to connect, they are then supposed to be double quoted. do you want to support that?
[source](https://www.firebirdsql.org/file/documentation/html/en/firebirddocs/qsg3/firebird-3-quickstartguide.html#qsg3-databases-connecting-isql)  (do a text search for "`funny stuff`")

If so, in addition to being careful about what is actually being matched, you will need some kind of escape or extra delimiters to disambiguate parses. For instance
```
parse(r#"firebird://a@::@@192.168.0.1//srv/db/database_name.fdb?dialect=3"#)
```
in this example, is (user,pass) meant to be `(a@:,@)` or `(a@,:@)` ?
(also i am not sure if double quotes themselves are supposed to be allowed, then there are additional edge cases if we support double quotes in usernames/passwords and also use them as delimiters)

if you prefer to just not support usernames/passwords containing `@` or `:`, `"` or other symbols for now, I'd consider that acceptable, so long as there is a doc comment about it.

Reply:
I think it is better to use the [`url`](https://docs.rs/url/2.2.0/url/index.html) crate to avoid parsing problems such as those ones.

Reply:
Ok, I didn't think about the passwords with '@'. I will try use the url crate to avoid this problems instead of try explain all the regex